src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs

```diff
@@ -77,27 +92,28 @@ private enum CompatibilityLevel
         // Tracks how often we meed to look for new data
         private DateTime _timeCurVersionDataLastUpdatedUtc = DateTime.MinValue;
         private VersionCompatibilityData _curVersionCompatibilityData;
+        private IVsSolution _vsSolution;
 
         public async Task InitializeAsync()
         {
             await _threadHandling.Value.SwitchToUIThread();
 
             // Initialize our cache file
-            string appDataFolder = await _shellUtilitiesHelper.Value.GetLocalAppDataFolderAsync(_serviceProvider);
+            string appDataFolder = await _shellUtilitiesHelper.Value.GetLocalAppDataFolderAsync(_vsShellService);
             if (appDataFolder != null)
             {
                 _versionDataCacheFile = new RemoteCacheFile(Path.Combine(appDataFolder, VersionDataFilename), VersionCompatibilityDownloadFwlink,
                                                             TimeSpan.FromHours(CacheFileValidHours), _fileSystem.Value, _httpClient);
             }
 
-            _ourVSVersion = await _shellUtilitiesHelper.Value.GetVSVersionAsync(_serviceProvider);
+            _ourVSVersion = await _shellUtilitiesHelper.Value.GetVSVersionAsync(_vsAppIdService);
 
-            IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
-            Verify.HResult(solution.AdviseSolutionEvents(this, out _solutionCookie));
+            _vsSolution = await _vsSolutionService.GetValueAsync();
+            Verify.HResult(_vsSolution.AdviseSolutionEvents(this, out _solutionCookie));
```

Comment:
There's a difference here that's worth noting. I'm caching the solution object so we don't have to call `GetService` in the non-async `Dispose` below. Any issue there?